Frameworks/CoreGraphics/CGContext.mm

```diff
@@ -17,296 +17,516 @@
+        return _impl.stateStack;
+    }
+
+    inline __CGContextDrawingState& CurrentGState() {
+        return GStateStack().top();
+    }
+
+    inline void ClearPath() {
+        _impl.currentPath.reset();
+    }
+
+    inline ComPtr<ID2D1Factory> Factory() {
+        ComPtr<ID2D1Factory> factory;
+        _impl.renderTarget->GetFactory(&factory);
+        return factory;
+    }
+};
 
-#define DEBUG_CONTEXT_COUNT
+#pragma region CFRuntimeClass
+static void __CGContextInit(CFTypeRef cf) {
+    CGContextRef context = (CGContextRef)cf;
+    struct __CGContext* mutableContext = const_cast<struct __CGContext*>(context);
+    new (&mutableContext->_impl) __CGContext::__CGContextImpl();
+}
 
-int contextCount = 0;
-static IWLazyClassLookup _LazyUIImage("UIImage"), _LazyUIScreen("UIScreen");
+static void __CGContextDeallocate(CFTypeRef cf) {
+    CGContextRef context = (CGContextRef)cf;
+    context->_impl.~__CGContextImpl();
+}
 
-pthread_mutex_t _cairoLock = PTHREAD_MUTEX_INITIALIZER;
+template <typename T, typename PT = typename std::remove_pointer<T>::type*, typename B = decltype(std::declval<PT>()->_base)>
+static PT _CFRuntimeCreateInstance(CFTypeID typeID, CFAllocatorRef allocator = kCFAllocatorDefault) {
+    return static_cast<PT>(
+        const_cast<void*>(_CFRuntimeCreateInstance(allocator, typeID, sizeof(typename std::remove_pointer<T>::type) - sizeof(B), nullptr)));
+}
 
-@interface CGNSContext : _CGLifetimeBridgingType
-@end
+/**
+ @Status Interoperable
+*/
+CFTypeID CGContextGetTypeID() {
+    static CFTypeID __kCGContextTypeID = _kCFRuntimeNotATypeID;
+    static const CFRuntimeClass __CGContextClass = { 0,       "CGContext", __CGContextInit, nullptr, __CGContextDeallocate,
+                                                     nullptr, nullptr,     nullptr,         nullptr };
 
-@implementation CGNSContext
-#pragma clang diagnostic push
-#pragma clang diagnostic ignored "-Wobjc-missing-super-calls"
-- (void)dealloc {
-    delete (__CGContext*)self;
+    static dispatch_once_t initOnce = 0;
+    dispatch_once(&initOnce, ^{
+        __kCGContextTypeID = _CFRuntimeRegisterClass(&__CGContextClass);
+    });
+    return __kCGContextTypeID;
 }
-#pragma clang diagnostic pop
-@end
+#pragma endregion
 
-__CGContext::__CGContext(CGImageRef pDest) {
-    contextCount++;
+static const wchar_t* TAG = L"CGContext";
 
-#ifdef DEBUG_CONTEXT_COUNT
-    TraceVerbose(TAG, L"contextCount: %d", contextCount);
-#endif
-    object_setClass((id) this, [CGNSContext class]);
-    scale = 1.0f;
-    _backing = pDest->Backing()->CreateDrawingContext(this);
+static void __CGContextInitializeWithRenderTarget(CGContextRef context, ID2D1RenderTarget* renderTarget) {
+    context->_impl.renderTarget = renderTarget;
+    CGContextSetRGBFillColor(context, 0, 0, 0, 0);
+    CGContextSetRGBStrokeColor(context, 0, 0, 0, 1);
 }
 
-__CGContext::~__CGContext() {
-    contextCount--;
+CGContextRef _CGContextCreateWithD2DRenderTarget(ID2D1RenderTarget* renderTarget) {
```

Comment:
The preceding function is called __CGContextInitializeWithRenderTarget whereas this one is _CGContextCreateWithD2DRenderTarget. Could we change one to make it more consistent with the other?